server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/RandomStoring.java

```diff
@@ -0,0 +1,201 @@
+    private final MailboxManager mailboxManager;
+    private final UserMailboxesService userMailboxesService;
+    private final MetricFactory metricFactory;
+    private final Iterator<Integer> randomRecipientsNumbers;
+    private MailDispatcher mailDispatcher;
+
+    @Inject
+    public RandomStoring(UsersRepository usersRepository, @Named("mailboxmanager") MailboxManager mailboxManager,
+                         UserMailboxesService userMailboxesService, MetricFactory metricFactory) {
+        this.metricFactory = metricFactory;
+        this.usersRepository = usersRepository;
+        this.mailboxManager = mailboxManager;
+        this.userMailboxesService = userMailboxesService;
+        this.randomRecipientsNumbers = new Random().ints(MIN_NUMBER_OF_RECIPIENTS, MAX_NUMBER_OF_RECIPIENTS + 1).boxed().iterator();
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        try {
+            rerouteMail(mail);
+            mailDispatcher.dispatch(mail);
+        } catch (UsersRepositoryException e) {
+            throw new MessagingException("Unable to compute a random user list", e);
+        }
+    }
+
+    private void rerouteMail(Mail mail) throws UsersRepositoryException {
+        Collection<ReroutingInfos> reroutingInfos = generateRandomMailboxes();
+        Collection<MailAddress> mailAddresses = reroutingInfos
+            .stream()
+            .map(ReroutingInfos::getMailAddress)
+            .collect(Guavate.toImmutableList());
+
+        mail.setRecipients(mailAddresses);
+        reroutingInfos.forEach(reroutingInfo ->
+            mail.setAttribute(Attribute.convertToAttribute(MailStore.DELIVERY_PATH_PREFIX + reroutingInfo.getUser(), reroutingInfo.getMailbox())));
+    }
+
+    @Override
+    public String getMailetInfo() {
+        return "Random Storing Mailet";
+    }
+
+    @Override
+    public void init() throws MessagingException {
+        mailDispatcher = MailDispatcher.builder()
+            .mailStore(SimpleMailStore.builder()
+                .mailboxAppender(new MailboxAppender(mailboxManager))
+                .usersRepository(usersRepository)
+                .folder(MailboxConstants.INBOX)
+                .metric(metricFactory.generate(RANDOM_STORING_MAILS_METRIC_NAME))
+                .build())
+            .consume(getInitParameter("consume", true))
+            .mailetContext(getMailetContext())
+            .build();
+    }
+
+    private Collection<ReroutingInfos> generateRandomMailboxes() throws UsersRepositoryException {
+        List<ReroutingInfos> reroutingInfos = Streams.stream(usersRepository
+            .list())
+            .map(User::fromUsername)
+            .flatMap(this::buildMailboxesForAnUser)
+            .distinct()
+            .collect(Collectors.toList());
+
+        Collections.shuffle(reroutingInfos);
+        return reroutingInfos
+            .stream()
+            .limit(randomRecipientsNumbers.next())
```

Comment:
you are limiting number of `reroutingInfos`, not number of recipients.

Reply:
How about put randomizing in 2 steps:
 - get random recipients
 - generates random ReRoutingInfor for those recipients

Reply:
Sure, I do not thing it matters for what we are trying to do.